src/Castle.Core/DynamicProxy/ModuleScope.cs

```diff
@@ -174,11 +189,94 @@ public Type GetFromCache(CacheKey key)
 		/// </summary>
 		/// <param name = "key">The key to be associated with the type.</param>
 		/// <param name = "type">The type to be stored in the cache.</param>
+		[Obsolete]
+		[EditorBrowsable(EditorBrowsableState.Never)]
 		public void RegisterInCache(CacheKey key, Type type)
 		{
 			typeCache[key] = type;
 		}
 
+		/// <summary>
+		///   Returns a type from this scope's type cache, or registers it in the cache if the key cannot be found.
+		/// </summary>
+		/// <param name="key">The key to be looked up in the cache.</param>
+		/// <param name="valueFactory">A function producing the type to be registered in the cache.</param>
+		/// <returns>The type from this scope's type cache matching the key, or null if the key cannot be found</returns>
+		/// <remarks>
+		///   This method does not synchronizes accesses to this scope's type cache using read/write locks.
+		///   Only use this method when you know for sure that a method further up in the call stack already holds
+		///   a write lock. The function must not access this scope's type cache.
+		/// </remarks>
+		public Type GetFromOrRegisterInCache(CacheKey key, Func<CacheKey, Type> valueFactory)
+		{
+			Type value;
+
+			Debug.Assert(key != null);
+			Debug.Assert(valueFactory != null);
+
+			if (this.typeCache.TryGetValue(key, out value))
```

Comment:
No need to prefix all field access with `this.`.

Reply:
Sorry, that's a (bad?) habit I picked up from a recent coworker. 😄 Will remove.